# Model Formulation 

- Discriminative Probability Classifier
- Models the probability of a binary outcome as a smooth function of predictors, constrained between 0 and 1.

$$
\Large\begin{align*}Y &\in \{0,1\}\\P(Y=1) &=p\\P(Y=0) &= 1-p\\P(Y=y) &= p^y(1-p)^{1-y}\\\end{align*}
$$

Why is log-odds the natural parameter of Bernoulli distribution?

- We start with Bernoulli pdf
- Write it in exponential form
- The coefficient multiply $y$ must be the natural parameter
- so log-odds is not a design choice, it emerges.

Consider; $p^y(1-p)^{1-y}$, lets write it in exponential form;

$$
\Large\begin{align*}p^y(1-p)^{1-y} &= exp(y\log{p} + (1-y) \log{1-p}))\\&= exp(y\log{p} + \log(1-p) - y\log{1-p})\\&= exp(y\log{\frac{p}{1-p}}+ \log{1-p})\\\end{align*}
$$

Canonical exponential form is as follows; 

$$
\Large\begin{align*}f(y|\theta) = exp(y\theta - A(\theta))\end{align*}
$$

So, based on our exponential form of Bernoulli, we have ; 

$$
\Large\begin{align*}\theta &= \log{\frac{p}{1-p}}\\-A(\theta) &= \log{1-p}\\\end{align*}
$$

So, $\log{\frac{p}{1-p}}$ is a natural parameter.

Introducing Covariates : Conditional Bernoulli

Once predictors exist, we are no longer modeling p, but ; 

$$
\Large\begin{align*}p(x) = P(Y=1|X=x)\end{align*}
$$

So, the natural parameter becomes conditional; 

$$
\Large\begin{align*}\theta(x) = \log{(\frac{p(x)}{1-p(x)})}\end{align*}
$$

We make one structural assumption; The natural parameter depends linearly on predictors; 

$$
\Large\begin{align*}\theta(x) = \beta_0 + \beta^Tx\end{align*}
$$

- $\theta \in (-\infty,+\infty)$
- Linear structure
- Linear natural parameter implies convex log-likelihood

When we plug back the theta value into $p(x)$;

# Linking Functions 

- Probit - Normal CDF
- Complementary log-log
- Cauchit
- Only logit is the canonical link for Bernoulli
- Canonical link → best statistical properties.

# Parameter Estimation using MLE 

For a single observation $(x_i, y_i)$ with $y_i \in \{0,1\}$, the Bernoulli probability mass function is:

$$
\Large\begin{align*}P(Y_i = y_i \mid X_i = x_i)&= p_i^{y_i}(1-p_i)^{1-y_i}\end{align*}
$$

$$
\Large\begin{align*}p_i = P(Y_i=1 \mid X_i=x_i)\end{align*}
$$

Assuming **conditional independence** of observations given $X$, the likelihood for the full dataset $\{(x_i,y_i)\}_{i=1}^n$ is the product of individual likelihoods:

$$
\Large\begin{align*}L(\beta)&= \prod_{i=1}^{n} p_i^{y_i}(1-p_i)^{1-y_i}\end{align*}
$$

Here, the probabilities $p_i$ are modeled using the logistic function:

$$
\Large\begin{align*}p_i&= \frac{1}{1 + \exp(-\eta_i)}\\\eta_i&= \beta_0 + \beta^T x_i\end{align*}
$$

Since the log function is monotonic, maximising the likelihood is equivalent to maximising the log-likelihood.

$$
\Large\begin{align*}\ell(\beta)&= \log L(\beta)\\&= \sum_{i=1}^{n} \left[ y_i \log(p_i) + (1-y_i)\log(1-p_i) \right]\end{align*}
$$

This is the **log-likelihood** of logistic regression.
In optimisation, it is common to convert a maximisation problem into a minimisation problem. We therefore define the **negative log-likelihood (NLL)**:

$$
\Large\begin{align*}- \sum_{i=1}^{n} \left[ y_i \log(p_i) + (1-y_i)\log(1-p_i) \right]\end{align*}
$$

The goal of logistic regression is; 

$$
\Large\begin{align*}\hat{\beta}= \arg\min_{\beta} \; \mathcal{L}(\beta)\end{align*}
$$

**Logistic Regression chooses $\beta$ so that the predicted probabilities $\sigma(\beta^Tx_i)$ match the observed labels as well as possible.** 

The negative log-likelihood above is known in machine learning as the **binary cross-entropy loss** (or log loss).
For a single observation, the loss is:

$$
\Large\begin{align*}\text{BCE}(y_i, p_i)&= - \left[ y_i \log(p_i) + (1-y_i)\log(1-p_i) \right]\end{align*}
$$

For the full dataset; 

$$
\Large\begin{align*}\text{BCE}(\beta)&= \sum_{i=1}^{n} \text{BCE}(y_i, p_i)\end{align*}
$$

- This loss arises **directly from maximum likelihood estimation**, not as a heuristic.
- Thus, minimising binary cross-entropy is equivalent to **maximum likelihood estimation for a Bernoulli model with a logistic link**.

Thus, Logistic regression estimates parameters $\beta$ by:

1. Modeling $P(Y=1 \mid X=x)$ via the logistic (sigmoid) function
2. Writing the Bernoulli likelihood for observed labels
3. Maximising the log-likelihood (or equivalently, minimising binary cross-entropy)

# Optimization Talks

Binary cross entropy loss comes. directly from MLE 

Logistic regression = maximise log-likelihood

Equivalently = minimise BCE Loss

The algorithm, it starts with random parameters \beta . The probabilities are computed. There are many algorithms which we can use for iterative optimisation ( logistic regression has no closed form solution)

Logistic regression does not have closed form solution. 

If we take the gradient and set to zero. The equation will be nonlinear in beta. Because sigmoid function contains an exponential. No algebraic manipulation can isolate beta. So, there is no analytic solution to that equation. 

The sigmoid introduces exponentials of $\beta$ making the likelihood equations transcendental rather than linear or polynomial.

# Canonical Exponential Form

A standardised way to write a probability distribution as an exponential of a linear function.

# Implementation

In [17]:
import numpy as np

class LogisticRegressionGD:
    def __init__(self, x, y, iterations=1000, learning_rate=0.01):
        """
        x: np.array of shape (n_samples, n_features)
        y: np.array of shape (n_samples, 1) with 0/1 labels
        """
        self.x = x
        self.y = y
        self.n_iter = iterations
        self.lr = learning_rate
        self.n_samples, self.n_features = x.shape

        # weights (column vector) and bias
        self.w = np.zeros((self.n_features, 1))
        self.b = 0

    def sigmoid(self, z):
        return 1 / (1 + np.exp(-z))

    def train(self, verbose=False):
        for i in range(1, self.n_iter + 1):
            # linear model
            z = self.x.dot(self.w) + self.b
            y_pred = self.sigmoid(z)

            # residuals / error
            residuals = y_pred - self.y  # shape (n_samples,1)

            # gradients
            dw = (1 / self.n_samples) * self.x.T.dot(residuals)
            db = np.mean(residuals)

            # update weights
            self.w -= self.lr * dw
            self.b -= self.lr * db

            if verbose and (i % 100 == 0 or i == 1):
                # cross-entropy loss
                eps = 1e-15  # to avoid log(0)
                loss = -np.mean(self.y * np.log(y_pred + eps) + (1 - self.y) * np.log(1 - y_pred + eps))
                print(f"Iteration {i}: Loss = {loss:.4f}, Weights = {self.w.flatten()}, Bias = {self.b:.4f}")

    def predict_proba(self, x_new):
        z = x_new.dot(self.w) + self.b
        return self.sigmoid(z)

    def predict(self, x_new, threshold=0.5):
        prob = self.predict_proba(x_new)
        return (prob >= threshold).astype(int)


In [18]:

np.random.seed(0)

# Sample data (10 samples, 2 features)
X = np.random.randn(10, 2)
y = (X[:,0] + X[:,1] > 0).astype(int).reshape(-1,1)  # simple linear separable labels

# Train logistic regression
model = LogisticRegressionGD(X, y, iterations=1000, learning_rate=0.1)
model.train(verbose=True)

# Predict
X_new = np.array([[0,0],[1,1],[-1,2]])
predictions = model.predict(X_new)
probs = model.predict_proba(X_new)

print("Predictions:\n", predictions)
print("Probabilities:\n", probs)


Iteration 1: Loss = 0.6931, Weights = [0.03993587 0.02240787], Bias = 0.0400
Iteration 100: Loss = 0.1921, Weights = [1.13718    1.01517886], Bias = 1.0452
Iteration 200: Loss = 0.1450, Weights = [1.45763373 1.60411047], Bias = 1.1500
Iteration 300: Loss = 0.1177, Weights = [1.6975981  2.06600644], Bias = 1.1682
Iteration 400: Loss = 0.0990, Weights = [1.90727415 2.44381287], Bias = 1.1720
Iteration 500: Loss = 0.0852, Weights = [2.09610572 2.76188257], Bias = 1.1742
Iteration 600: Loss = 0.0747, Weights = [2.26742613 3.03595227], Bias = 1.1768
Iteration 700: Loss = 0.0665, Weights = [2.4235189  3.27650733], Bias = 1.1802
Iteration 800: Loss = 0.0598, Weights = [2.56635126 3.49076495], Bias = 1.1841
Iteration 900: Loss = 0.0544, Weights = [2.69764648 3.68386545], Bias = 1.1886
Iteration 1000: Loss = 0.0498, Weights = [2.81889003 3.85958716], Bias = 1.1933
Predictions:
 [[1]
 [1]
 [1]]
Probabilities:
 [[0.76732981]
 [0.99961879]
 [0.99774777]]


In [20]:
np.exp(model.w),np.exp(model.b)

(array([[16.75823926],
        [47.44576002]]),
 np.float64(3.2979292708102528))

**Interpretations**
 - $\beta_1$ - change in log-odds for 1 unit increase in $x_1$
 - A one -unit increase in $x_1$ multiplies the odds of $y=1$ by 16.75 , holding other variables constant. 

# Key-Logistic Assumptions
 - Logit is linear in parameters 
 $$
 \Large \log\left(\frac{p}{1-p}\right) = X\beta
 $$ 
 - Observations are independent
 - No severe multicollinearity
 - No perfect separation

In [22]:
import numpy as np
import statsmodels.api as sm

np.random.seed(42)

# Number of samples
n = 100

# Generate 2 features
X1 = np.random.randn(n)
X2 = np.random.randn(n)

# Linear combination + noise
linear_comb = 0.5*X1 + 1.2*X2 - 0.3

# Apply sigmoid to get probabilities
prob = 1 / (1 + np.exp(-linear_comb))

# Generate binary labels
y = np.random.binomial(1, prob, size=n)

# Combine features
X = np.column_stack((X1, X2))

# Split into train/test (simple 80/20 split)
train_size = int(0.8 * n)
x_train, x_test = X[:train_size], X[train_size:]
y_train, y_test = y[:train_size], y[train_size:]

# Add constant for intercept
X_sm = sm.add_constant(x_train)

# Fit logistic regression
logit_model = sm.Logit(y_train, X_sm)
result = logit_model.fit()

# Print summary
print(result.summary())


Optimization terminated successfully.
         Current function value: 0.485725
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:                      y   No. Observations:                   80
Model:                          Logit   Df Residuals:                       77
Method:                           MLE   Df Model:                            2
Date:                Sun, 01 Feb 2026   Pseudo R-squ.:                  0.2992
Time:                        16:38:12   Log-Likelihood:                -38.858
converged:                       True   LL-Null:                       -55.452
Covariance Type:            nonrobust   LLR p-value:                 6.214e-08
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.0211      0.281     -0.075      0.940      -0.572       0.530
x1             0.6451      0.

In [23]:
params = result.params
conf = result.conf_int()
odds_ratios = np.exp(params)
conf_odds = np.exp(conf)

print(params)
print(conf)
print(odds_ratios)
print(conf_odds)

[-0.02106762  0.64510975  1.75547306]
[[-0.57220158  0.53006635]
 [ 0.01676804  1.27345147]
 [ 0.94898477  2.56196135]]
[0.97915276 1.90619622 5.78618433]
[[ 0.56428176  1.69904504]
 [ 1.01690941  3.57316396]
 [ 2.58308591 12.96121394]]


# Odds, log -odds, odds ratio - Interpretations 

$$
\Large odds = \frac{p}{1-p}
$$ 
* Odds gives - success ( p ) is how many times as likely as failure.
* This value ranges from 0 to infinity, can't model linearly with unrestricted Xβ
* When we take the log of odds, this ranges from -infinity to + inifity.
* So +log odds means probability >0.5 and -ve log odds means probability <0.5
* In logistic regression we assume the following; The log odds has has a linear parameter structure with the predictors.This value can be modelled linearly with unrestricted Xβ

$$ 
\Large \log(\frac{p}{1-p}) = \beta_0 + \beta_1 x_1 + ...
$$

This ensures that we can get the probability between 0 and 1 ( using sigmoid )

$$
\Large \log(\frac{p}{1-p}) = -1.05 + 1.95 x_1 -0.98 x_2
$$

* If Odds Ratios of the parameters are : 0.35, 7.0 and 0.37 respectively
* One unit increase in $x_1$ multiplies the odds of y=1 by 7
* One unit increase in $x_2$ reduces odds by 63% ( 1-0.37).
* 0.35 is the odds of y=1 at $x_1 = x_2 = 0 $

In [24]:
from sklearn.linear_model import LogisticRegression

sk_model = LogisticRegression(fit_intercept=True, solver="lbfgs")
sk_model.fit(x_train, y_train)

sk_model.intercept_, sk_model.coef_


(array([-0.02322445]), array([[0.54384023, 1.5090275 ]]))

# Overall Model Significance - Likelihood Ratio Test

In [25]:
import statsmodels.api as sm
from scipy import stats

X_sm = sm.add_constant(x_train)

# Null model (intercept only)
model_null = sm.Logit(y_train, np.ones((len(y_train), 1))).fit(disp=False)

# Full model
model_full = sm.Logit(y_train, X_sm).fit(disp=False)

LR_stat = 2 * (model_full.llf - model_null.llf)
df = X_sm.shape[1] - 1
p_value = stats.chi2.sf(LR_stat, df)

LR_stat, p_value


(np.float64(33.1876282883203), np.float64(6.214383967598268e-08))

# Wald Test - Single Coefficient Significance Test

In [26]:
import numpy as np
from scipy import stats


beta_hat = result.params
se = result.bse

# Wald test for coefficient j (example: x1 → index 1)
j = 1

z_stat = beta_hat[j] / se[j]
p_value = 2 * (1 - stats.norm.cdf(abs(z_stat)))

z_stat, p_value


(np.float64(2.012267921478266), np.float64(0.04419170202581291))

# Comparison

    
Aspect                             | Linear Regression                 | Logistic Regression                           |
| :-------------------------------- | :-------------------------------- | :-------------------------------------------- |
| **Linear assumption**              | (y) linear in (X)                 | **Log-odds** linear in (X)                    |
| **Error distribution**             | Normal                            | Binomial variance (p(1-p))                    |
| **Variance of response**           | Constant ((\sigma^2))             | Mean-dependent                                |
| **Estimation method**              | Least Squares                     | Maximum Likelihood                            |
| **Overall model test**             | ANOVA F-test                      | Likelihood Ratio Test (LRT)                   |
| **Individual parameter test**      | t-test                            | Wald test                                     |
| **Alternative tests**              | Partial F-test                    | Wald / LRT / Score                            |
| **Model comparison**               | F-test                            | Likelihood ratio                              |
| **Test statistic distribution**    | F, t                              | $( \chi^2 )$, Normal (asymptotic)              |
| **Confidence intervals**           | Exact (normality)                 | Asymptotic (Wald / profile)                  |
| **Interpretation of coefficients** | Change in mean (y)                | Change in **log-odds**                        |
| **Natural effect scale**           | Units of (y)                      | Odds ratios                                   |
| **Exponentiated coefficients**     | Meaningless                        | Odds ratios                                   |
| **Residuals**                      | Raw, standardized                 | Deviance, Pearson                             |
| **Residual patterns**              | Homoscedastic                     | Mean–variance linked                          |
| **Pseudo-R²**                      | Not needed                         | McFadden, Cox–Snell                           |
| **Influence measures**             | Cook’s distance                   | Leverage, ΔDeviance                           |
| **Perfect separation**             | Not an issue                      | Can break MLE                                 |
| **Multicollinearity**              | Inflates SEs                      | Inflates SEs                                  |
